server/schedule/placement/rule_manager.go

```diff
@@ -659,3 +664,11 @@ func (m *RuleManager) SetKeyType(h string) *RuleManager {
 	m.keyType = h
 	return m
 }
+
+// SetStores will update stores info for adjustRule()
+func (m *RuleManager) SetStores(h []*core.StoreInfo) *RuleManager {
```

Comment:
Its easy to forget `SetStores` before `SetRules`. I suggest passing in a closure to get stores when creating rule management. like: 
```
c.ruleManager = placement.NewRuleManager(c.storage, c.GetStores)
```

Reply:
Nice!